File: samples/java/src/main/java/com/agents/patterns/coordination/GroupChat.java
package com.agents.patterns.coordination;
import org.springframework.ai.chat.client.ChatClient;
import java.util.*;
public class GroupChat {
    public record Turn(String speaker, String text) {}
    @FunctionalInterface public interface Manager { String next(List<Turn> thread, List<String> names); }
    private final Manager manager;
    private final int maxTurns;
    private final Map<String, java.util.function.Function<List<Turn>, String>> participants = new LinkedHashMap<>();
    public GroupChat(Manager manager, int maxTurns) { this.manager = manager; this.maxTurns = maxTurns; }
    public GroupChat add(String name, java.util.function.Function<List<Turn>, String> agent) { participants.put(name, agent); return this; }
    public List<Turn> run(String opening) {
        List<Turn> thread = new ArrayList<>();
        thread.add(new Turn("input", opening));
        List<String> names = new ArrayList<>(participants.keySet());
        for (int i = 0; i < maxTurns; i++) {
            String speaker = manager.next(thread, names);
            if (speaker == null) return thread;
            String msg = participants.get(speaker).apply(thread);
            thread.add(new Turn(speaker, msg));
        }
        return thread;
    }
    public static Manager makerCheckerManager(String approveToken) {
        return (thread, names) -> {
            for (int i = thread.size() - 1; i >= 0; i--) {
                Turn t = thread.get(i);
                if (t.speaker().equals("checker")) {
                    if (t.text().contains(approveToken)) return null;
                    break;
                }
            }
            String last = thread.get(thread.size() - 1).speaker();
            return last.equals("maker") ? "checker" : "maker";
        };
    }
    public static String runWithLlm(ChatClient chatClient, String goal) {
        java.util.function.Function<List<Turn>, String> maker = thread -> {
            long drafts = thread.stream().filter(t -> t.speaker().equals("maker")).count();
            String feedback = thread.stream().filter(t -> t.speaker().equals("checker")).reduce((a, b) -> b).map(Turn::text).orElse("none yet");
            return chatClient.prompt().user("Draft #" + (drafts + 1) + ". Feedback: " + feedback + ". Write a 2-sentence draft for: " + goal).call().content();
        };
        java.util.function.Function<List<Turn>, String> checker = thread -> {
            String draft = thread.stream().filter(t -> t.speaker().equals("maker")).reduce((a, b) -> b).map(Turn::text).orElse("");
            return chatClient.prompt().user("Review draft: " + draft + ". If acceptable, start with APPROVED. Otherwise REJECTED with one line feedback.").call().content();
        };
        GroupChat chat = new GroupChat(makerCheckerManager("APPROVED"), 8).add("maker", maker).add("checker", checker);
        return chat.run(goal).stream().map(t -> "[" + t.speaker() + "] " + t.text()).reduce("", (a, b) -> a + "\n" + b);
    }
}

File: samples/csharp/src/AgentPatterns/Coordination/GroupChat.cs
using Microsoft.SemanticKernel.ChatCompletion;
namespace AgentPatterns.Coordination;
public record Turn(string Speaker, string Text);
public class GroupChat {
    public delegate string? Manager(IList<Turn> thread, IList<string> names);
    private readonly Manager _manager;
    private readonly int _maxTurns;
    private readonly Dictionary<string, Func<IList<Turn>, string>> _participants = new();
    public GroupChat(Manager manager, int maxTurns = 10) { _manager = manager; _maxTurns = maxTurns; }
    public GroupChat Add(string name, Func<IList<Turn>, string> agent) { _participants[name] = agent; return this; }
    public IList<Turn> Run(string opening) {
        var thread = new List<Turn> { new("input", opening) };
        var names = _participants.Keys.ToList();
        for (int i = 0; i < _maxTurns; i++) {
            var speaker = _manager(thread, names);
            if (speaker is null) return thread;
            thread.Add(new Turn(speaker, _participants[speaker](thread)));
        }
        return thread;
    }
    public static Manager MakerCheckerManager(string approveToken = "APPROVED") => (thread, _) => {
        for (int i = thread.Count - 1; i >= 0; i--) {
            if (thread[i].Speaker == "checker") {
                if (thread[i].Text.Contains(approveToken)) return null;
                break;
            }
        }
        return thread[^1].Speaker == "maker" ? "checker" : "maker";
    };
    public static async Task<IList<Turn>> RunWithLlm(IChatCompletionService svc, string goal) {
        async Task<string> Invoke(string prompt) {
            var h = new ChatHistory(prompt);
            return (await svc.GetChatMessageContentsAsync(h))[0].Content ?? "";
        }
        int drafts = 0;
        Func<IList<Turn>, string> maker = thread => {
            drafts++;
            var feedback = thread.LastOrDefault(t => t.Speaker == "checker")?.Text ?? "none yet";
            return Invoke($"Draft #{drafts}. Feedback: {feedback}. Write a 2-sentence draft for: {goal}").GetAwaiter().GetResult();
        };
        Func<IList<Turn>, string> checker = thread => {
            var draft = thread.LastOrDefault(t => t.Speaker == "maker")?.Text ?? "";
            return Invoke($"Review: {draft}. If good, start with APPROVED. Otherwise REJECTED with one line.").GetAwaiter().GetResult();
        };
        return new GroupChat(MakerCheckerManager(), 8).Add("maker", maker).Add("checker", checker).Run(goal);
    }
}
